operators/src/upgrade/k8s/crd/v0.rs

```diff
@@ -0,0 +1,223 @@
+    }
+
+    /// Returns the desired version of the product
+    pub fn target_version(&self)->Option<Version>{
+        self.target_version.as_ref().map(|v| Version::from_str(v).unwrap())
+    }
+
+    /// Returns all the components
+    pub fn components(&self)->Option<HashMap<String,Vec<String>>>{
+        self.components.clone()
+    }
+}
+
+
+/// 'UpgradePhase' defines the status of each components
+#[derive(Deserialize, Serialize, Clone, Debug, Eq, PartialEq, JsonSchema)]
+pub enum UpgradePhase{
+    Waiting,
+
+    Updating,
+
+    Verifying,
+
+    Completed,
+
+    Error,
+}
+
+impl Default for UpgradePhase{
+    fn default() -> Self {
+        UpgradePhase::Waiting
+    }
+}
+
+
+/// converts the Upgrade phase into a string
+impl ToString for UpgradePhase {
+    fn to_string(&self) -> String {
+        match self {
+            UpgradePhase::Waiting => "Waiting",
+            UpgradePhase::Updating => "Updating",
+            UpgradePhase::Verifying => "Verifying",
+            UpgradePhase::Completed => "Completed",
+            UpgradePhase::Error => "Error",
+        }
+        .to_string()
+    }
+}
+/// Upgrade phase into a string
+impl From<UpgradePhase> for String {
+    fn from(u: UpgradePhase) -> Self {
+        u.to_string()
+    }
+}
+
+/// 'UpgradeConditionType' defines the status of upgradeaction resource
+#[derive(Deserialize, Serialize, Clone, Debug,  Eq, PartialEq, JsonSchema)]
+pub enum UpgradeConditionType{
+
+    NotStarted,
+
+    Updating,
+
+    VerifyingUpdate,
+
+    SuccessfullUpdate,
+
+    Error,
+
+}
+
+impl Default for UpgradeConditionType{
+    fn default() -> Self {
+        UpgradeConditionType::NotStarted
+    }
+}
+
+/// converts the Upgrade Condition Type into a string
+impl ToString for UpgradeConditionType {
+    fn to_string(&self) -> String {
+        match self {
+            UpgradeConditionType::NotStarted => "NotStarted",
+            UpgradeConditionType::Updating => "Updating",
+            UpgradeConditionType::VerifyingUpdate => "VerifyingUpdate",
+            UpgradeConditionType::SuccessfullUpdate => "SuccessfullUpdate",
+            UpgradeConditionType::Error => "Error",
+        }
+        .to_string()
+    }
+}
+/// Upgrade Condition Type into a string
+impl From<UpgradeConditionType> for String {
+    fn from(u: UpgradeConditionType) -> Self {
+        u.to_string()
+    }
+}
+
+#[derive(Deserialize, Serialize, Clone, Debug,Eq, PartialEq, JsonSchema)]
+pub struct UpgradeCondition{
+    /// Type of upgrade condition
+    #[validate(required)]
+    pub type_: Option<UpgradeConditionType>,
+
+    /// Last time the condition transit from one status to another.
+    pub last_transition_time: Option<String>,
+
+    ///Human readable message indicating details about last transition.
+    pub message: Option<String>,
+
+    ///(one line) reason for the condition's last transition.
+    pub reason: Option<String>,
+
+    /// Status of the cpondition, one of True, False, Unknown.
```

Comment:
Controller updates the condition and when a new CR is created with all the component details, default condition is going to be created. Conditions and upgrade phases are inter linked to each other.
But instead of using conditions, I am thinking of using state and last_transition_time in UpgradeActionStatus. it would also work and can make our work easier.

Reply:
Could replace String with an enum, example:
enum ConditionStatus { Unknown, True, False } ?

Reply:
> Controller updates the condition and when a new CR is created with all the component details, default condition is going to be created. Conditions and upgrade phases are inter linked to each other. But instead of using conditions, I am thinking of using state and last_transition_time in UpgradeActionStatus. it would also work and can make our work easier.

Yeah. Otherwise maintaining conditions will be an overhead and can generate wrong notions if not maintained properly.